src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs

```diff
@@ -464,15 +466,34 @@ public virtual int TupleElementIndex
             get
             {
                 // wrapped tuple fields already have this information and override this property
-                Debug.Assert(!(this is TupleElementFieldSymbol or TupleErrorFieldSymbol));
+                Debug.Assert(!(this is TupleElementFieldSymbol or TupleErrorFieldSymbol or Retargeting.RetargetingFieldSymbol));
+                if (!ContainingType.IsTupleType)
+                {
+                    return -1;
+                }
 
-                var map = ContainingType.TupleFieldDefinitionsToIndexMap;
-                if (map is object && map.TryGetValue(this.OriginalDefinition, out int index))
+                if (!ContainingType.IsDefinition)
                 {
-                    return index;
+                    return this.OriginalDefinition.TupleElementIndex;
                 }
 
-                return -1;
+                var tupleElementPosition = NamedTypeSymbol.MatchesCanonicalTupleElementName(Name);
+                int arity = ContainingType.Arity;
+                if (tupleElementPosition <= 0 || tupleElementPosition > arity)
+                {
+                    // ex: no "Item2" in 'ValueTuple<T1>'
+                    return -1;
+                }
+                Debug.Assert(tupleElementPosition < NamedTypeSymbol.ValueTupleRestPosition);
+
+                WellKnownMember wellKnownMember = NamedTypeSymbol.GetTupleTypeMember(arity, tupleElementPosition);
+                MemberDescriptor descriptor = WellKnownMembers.GetDescriptor(wellKnownMember);
+                Symbol found = CSharpCompilation.GetRuntimeMember(ImmutableArray.Create<Symbol>(this), descriptor, CSharpCompilation.SpecialMembersSignatureComparer.Instance,
+                    accessWithinOpt: null); // force lookup of public members only
+
+                return found is not null
+                    ? tupleElementPosition - 1
+                    : -1;
```

Comment:
is any of this expensive?  would it be worth caching the reuslt of this?

Reply:
Considered this question and I lean towards not caching. 
When using tuple syntax, we create `TupleElementFieldSymbol` for default elements (and `TupleVirtualElementFieldSymbol` for named elements) which stores location and name. So this path is only used when accessing a field directly from `ValueTuple` type (from source or PE) and that'd be less common.
I've filed an [issue](https://github.com/dotnet/roslyn/issues/58499) for discussion which proposes to remove this wrapping with `TupleElementFieldSymbol`. If we go down that route then caching would make more sense.